### DISCUSSION TITLE
Mirai Re-design 2

### DISCUSSION THREAD

**Discussion by shikokuchuo** [r-lib/mirai#46] [Ideas]
Just so you hear it here first. It is likely possible to eliminate the dispatcher altogether, just through use of the condition variables.

I still need to gather my thoughts on this.

:)

@wlandau

**Comment by wlandau** [comment]
Awesome! I would love to move beyond the dispatcher process if possible.

**Comment by shikokuchuo** [comment]
My initial idea here was just using a load-balancing algorithm - send tasks to servers which are *currently* not busy. Of course that doesn't help if all of them are busy and we don't know how long the tasks will last.

This is not optimal vs. the dispatcher, but better than the non-dispatcher round-robin behaviour.

I want to avoid just replicating dispatcher on a thread to get real benefits. Even using a lightweight inproc transport to send/recv messages between threads doesn't realise the full potential of a threaded approach.

The NNG mutex locks are going to be key here... along with special accessor functions to the thread that uses them so everything remains thread safe. This implies that there will be a layer that runs entirely at the C level, inaccessible to R so that everything is kept safe.

The new `nanonext::cv_signal()` will be the key enabler.

**Reply by wlandau** [reply]
Interesting. Is the thought to implement this kind of passive load-balancing in the `nanonext` C code? In what way is it suboptimal compared to the dispatcher? Could it be expensive to find a free daemon for each new task?

**Reply by shikokuchuo** [reply]
It's good when the load is light, or it helps obvious mistakes so that tasks don't queue up behind just one backlogged daemon when others are free. But if you think about it, under heavy load, all daemons are not free and then the algorithm is no better.

Passive load-balancing doesn't seem to be the way forward as you'd want tasks to be submitted when daemons free up - and this could be any time in the future. You need another thread or process to wait on these events as they happen. The only alternative seems to be suboptimal polling using an event loop as in `later` - which can also only execute when nothing else is happening.

**Reply by wlandau** [reply]
So this is more like the threaded dispatcher option, as opposed to the req/rep 1.0 idea from https://github.com/nanomsg/nng/issues/10 of "letting [daemon] pipes race to get the [host] pipe"? I was hoping something like the latter might somehow be possible. Maybe it's not always strictly faster than active dispatch, but it fits conceptually with `crew`'s model and could really simplify it at all levels.

**Reply by shikokuchuo** [reply]
I’d be interested to know how this could simplify things at {crew}. If you want this kind of behaviour, we could just swap the req/rep relationship between dispatcher and daemons. A daemon will send a message I want a task, gets sent one, then sends the reply, dispatcher sends an acknowledgment . And repeat. This requires 4 messages rather than 2 so is less efficient, but the first and last can just be a nul byte or sth. This would solve the exit linger period as well. But other benefits?

**Reply by wlandau** [reply]
That's really interesting. On reflection, if `crew` is still integrated with `mirai`, maybe dropping the exit linger period would be the only benefit. I guess I was thinking more in terms of a direct `crew`/`nanonext` integration. With `crew`, I was hoping for a passive dispatch messaging model like this:

1. The host passively/asynchronously pushes tasks to a single central listener socket.
2. The hosts receives tasks from a single central listener socket, not necessarily the same as (1).
3. Daemons dial into (1) to get tasks on a first-come-first-served basis (no round-robin).
4. Daemons push results to (2).

If I understand correctly, this would eliminate the need for a tasks-only dispatcher, task-specific aio objects (which are cumbersome to process and difficult to whip through efficiently), different URLs for different daemons, URL token rotation, checks on `mirai` daemon status counters (especially for https://github.com/wlandau/crew/issues/79), and possibly https://github.com/wlandau/crew/issues/99 (if (1) and (2) are the only URLs (even better, if they can use the same listener/URL) then I can use direct TCP URLs instead of websockets and TLS will work better).

There may be more advantages that I am not thinking of right now. `mirai` was originally designed with a `future`-like model with handler objects for individual tasks, whereas `crew` just wants a single handler object to push and pop tasks. I would have to think about how auto-scaling would work in this new model, so that may complicate things a little bit, but on the whole the messaging model and internal design would align more with what `crew` is really trying to do. (The user interface would not need to change at all, and I think I could keep the launcher plugin interface back-compatible.)

But anyway, this is all what I was trying to figure out with my experiments at https://github.com/wlandau/crew/tree/prototypes/tests/prototypes. If you know of a way to accomplish this using req/rep in NNG/`nanonext`, I would be super eager to learn.

I plan on temporarily disabling interactions on GitHub later today so I can take a true break next week, but you know where to find me by email.

**Reply by shikokuchuo** [reply]
I think it's nice to think about conceptually, but actually trying to implement it, we would get back to something that looks a lot like what we have currently!

Some initial thoughts:
1. The central listener sockets will need to be on another thread (very difficult) or process ( = dispatcher). Passive won't work as you need forwarding to daemons at the minimum otherwise it would be the same as `mirai` without dispatcher.
2. Sending to one socket and retrieving from another (or the same) without being task-specific risks sending into the ether. We have to take it on trust we're going to get all our results back. If something crashes it would be lost and we'd be waiting indefinitely.
3. (3) With req/rep we're stuck with round robin for now.

It's always tempting to want to simplify, but let's let things bed down for a while with all the latest improvements.

Enjoy your vacation!

**Comment by shikokuchuo** [comment]
Superseded by #74
